### DISCUSSION TITLE
Observation Tokens and Steps

### DISCUSSION THREAD

**Discussion by beckydvn** [AI-Planning/macq#139] [General]
Hi there,

My question concerns `Observation` tokens - all of them take a `Step` as a parameter, and this is even expected in functions like the `tokenize` function in `Trace` (which ideally should be able to convert the steps in a trace to any type of `Observation` Token). This makes sense and so far has also allowed for some nice inheritance structures between the `Observation` classes.

The problem arises with the AMDN tokens, which I'm currently constructing. These can't take a `Step` as a parameter because `Step` doesn't have the capacity to hold the information the tokens need (in this case, sets of actions instead of a single action). Obviously modifying the ground truth `Step` is out of the question (I also imagine this problem could also arise again in the future as the extraction techniques continue to get more complex and the tokens require different things, so it would also only be a temporary hack).

I don't think we want to sacrifice having `Step` as a unified parameter for `Observation` token types, because:
- It makes inheritance easy
- We can convert traces to any token type
- There's been talk of converting token types to other token types (?)

But that leaves the question of what to do about AMDN tokens.

What do you think?

**Comment by haz** [comment]
Oh AMDN...so either a single token captures a bunch of actions, or it doesn't. If it does, then we need to modify it so that the token creation takes those multiple steps. Given that the tokenization has to happen at a higher level anyway (trace level and not on the individual steps), perhaps this is ok?

Only other option would be for the observation token to be associated with a single action, and have the extra info in the observation that indicates what parallel action it's a part of. Less-than-ideal I reckon...

**Reply by beckydvn** [reply]
Hmm ok, I suppose using multiple `Steps` for each token is a viable option. It's a bit redundant as they'll all have the same state, but at least they're conforming to the parameter structure of the rest of the `Observation` types. I suppose the mandatory `Step` would refer to the initial state and action, and then there would be an optional parameter for additional parallel actions... 

Second option works too, and might take a bit more finagling to work with, but might end up being clearer...

Also, as you mentioned - with the solution we came up with during the MAcq meeting, proper tokenization is happening at the trace level. Which is all good, unless someone calls `trace.tokenize` in which case we get invalid tokens because it was called at the step level. That is, there's no way to _enforce_ that it only happens at the trace level (unless we brute force check the token type and give the user a warning telling them to use a tracelist instead). So that's something else to consider structurally before we decide definitely...

**Reply by haz** [reply]
Well I think your final solution is the right one -- you throw an error if someone creates this new token type with a single action rather than a set of actions. The `__init__` will be overloaded, and you can swap out your assumptions on the parameters of that.

This all leads us to a unique type of tokenization, and raises the question if anything other than AMDN will use the same (set of actions + state). Even if the answer is "no", I think the collateral damage of needing a trace-level tokenization, etc, will still be useful and used again.
